### PR TITLE
[MIST-1114] Recover from master failure

### LINKED ISSUE #1114: Recovery from master failure
We need to recover MIST master when the failure occurs.

### PR DESCRIPTION
**Pull request by DifferentSC**
This PR resolves #1114 via

* Implement recovery process from master failures.

### CODE REVIEW

File: mist-core/src/main/java/edu/snu/mist/core/driver/MistDriver.java [hunk 1]
```diff
@@ -305,11 +330,25 @@ public void onNext(final FailedEvaluator failedEvaluator) {
...
       final InetSocketAddress failedNodeInetSocketAddress = failedEvaluator.getFailedContextList().get(0)
           .getEvaluatorDescriptor().getNodeDescriptor().getInetSocketAddress();
-      try {
-        proxyToMaster.notifyFailedTask(failedNodeInetSocketAddress.getHostName());
-      } catch (final AvroRemoteException e) {
-        LOG.log(Level.SEVERE, "Cannot connect to MistMaster for notifying failure! " + e.toString());
-        throw new IllegalStateException("Cannot connect to MistMaster while failure recovery");
+      if (failedEvaluator.getFailedContextList().get(0).getId().startsWith(MIST_MASTER_ID_PREFIX)) {
+        // Master has failed. We need to recovery master.
+        isMasterRunning.set(false);
```
[the remark this answers is not in the document]
Comment (DifferentSC) [reply mist-core/src/main/java/edu/snu/mist/core/driver/MistDriver.java]: In that case the worker is lost in the current implementation. I've created a separate issue in #1118.

File: mist-core/src/main/java/edu/snu/mist/core/driver/MistDriver.java [hunk 2]
```diff
@@ -305,11 +330,25 @@ public void onNext(final FailedEvaluator failedEvaluator) {
...
+        hasMasterFailed.set(true);
```
[the remark this answers is not in the document]
Comment (DifferentSC) [reply mist-core/src/main/java/edu/snu/mist/core/driver/MistDriver.java]: It does not become `false` until the system terminates.

File: mist-core/src/main/java/edu/snu/mist/core/master/DefaultApplicationCodeManager.java [hunk 3]
```diff
@@ -63,13 +64,28 @@
...
+
   @Inject
   private DefaultApplicationCodeManager(
-      @Parameter(SharedStorePath.class) final String jarStoringPath
-  ) {
+      @Parameter(SharedStorePath.class) final String jarStoringPath,
+      final ApplicationIdGenerator applicationIdGenerator,
+      @Parameter(DriverHostname.class) final String driverHostname,
+      @Parameter(MasterToDriverPort.class) final int masterToDriverPort
+  ) throws IOException {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/DefaultApplicationCodeManager.java]: please remove the newline

File: mist-core/src/main/java/edu/snu/mist/core/master/DefaultApplicationIdGenerator.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * The default application id generator.
+ */
+public final class DefaultApplicationIdGenerator implements ApplicationIdGenerator {
+
+  /**
+   * The atomic integer used for generating a unique app id.
+   */
+  private AtomicInteger appIndex;
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/DefaultApplicationIdGenerator.java]: final

File: mist-core/src/main/java/edu/snu/mist/core/master/DefaultApplicationIdGenerator.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  /**
+   * The index for the mist master.
+   */
+  private int masterIndex;
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/DefaultApplicationIdGenerator.java]: final

File: mist-core/src/main/java/edu/snu/mist/core/master/MistMaster.java [hunk 6]
```diff
@@ -60,39 +58,38 @@
...
-
   private final int initialTaskNum;
 
   private final TaskRequestor taskRequestor;
 
-  private final TaskStatsMap taskStatsMap;
+  private final MasterSetupFinished masterSetupFinished;
 
-  private final ProxyToTaskMap proxyToTaskMap;
+  private boolean hasMasterFailed;
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/MistMaster.java]: hasMasterFailed -> masterRecovery

File: mist-core/src/main/java/edu/snu/mist/core/master/MistMaster.java [hunk 7]
```diff
@@ -60,39 +58,38 @@
...
       final DriverToMasterMessage driverToMasterMessage,
       final ClientToMasterMessage clientToMasterMessage,
       final TaskToMasterMessage taskToMasterMessage,
       final TaskRequestor taskRequestor,
-      final TaskStatsMap taskStatsMap,
-      final ProxyToTaskMap proxyToTaskMap,
-      final MasterSetupFinished masterSetupFinished) {
-    this.masterToTaskPort = masterToTaskPort;
+      final MasterSetupFinished masterSetupFinished,
+      @Parameter(IsMasterFailed.class) final boolean hasMasterFailed,
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/MistMaster.java]: IsMasterFailed -> MasterRecovery ?

**Comment by taegeonum** [comment]
LGTM. I'm merging it. Thanks for the work!

**Pull request merged**